tasks/Task.cpp

```diff
@@ -2,48 +2,496 @@
+    {
+        iodrivers_base::RawPacket raw_packet;
+        if (_data_in.read(raw_packet) != RTT::NewData)
+            return;
+
+        vector<byte> data;
+        data.resize(raw_packet.data.size());
+        for (unsigned int i = 0; i < raw_packet.data.size(); i++)
+        {
+            data[i] = static_cast<byte>(raw_packet.data[i]);
+        }
+        mDataChannel->send(&data.front(), data.size());
+        break;
+    }
+    case DataChannelClosed:
+        throw std::runtime_error("DataChannel closed");
+    case DataChannelFailed:
+        throw std::runtime_error("DataChannel failed");
+    default:
+        break;
+    }
 }
 
-Task::~Task()
+void Task::evaluateWebSocket()
 {
+    switch (mState.web_socket)
+    {
+    case WebSocketClosed:
+        throw std::runtime_error("WebSocket closed");
+    case WebSocketFailed:
+        throw std::runtime_error("WebSocket failed");
+    default:
+        break;
+    }
 }
 
+Task::Task(string const &name) : TaskBase(name) {}
 
+Task::~Task() {}
 
 /// The following lines are template definitions for the various state machine
 // hooks defined by Orocos::RTT. See Task.hpp for more detailed
 // documentation about them.
 
 bool Task::configureHook()
 {
-    if (! TaskBase::configureHook())
+    if (!TaskBase::configureHook())
         return false;
+
+    mConfig.iceServers.emplace_back(_stun_server.get());
+    mWs = std::make_shared<rtc::WebSocket>();
+
+    configureWebSocket();
+
     return true;
 }
 bool Task::startHook()
 {
-    if (! TaskBase::startHook())
+    if (!TaskBase::startHook())
         return false;
+
+    if (!_remote_peer_id.get().empty())
+    {
+        std::future<void> wait_remote_peer_future = mWaitRemotePeerPromise.get_future();
+        mRemotePeerID = _remote_peer_id.get();
+        mRemotePeerAnswerReceived = false;
+
+        base::Time start_time = base::Time::now();
+        // Get contact with the remote peer and create datachannel
+        while (!mRemotePeerAnswerReceived)
+        {
+            base::Time duration = base::Time::now() - start_time;
+            if (duration > _wait_remote_peer_time_out.get())
+            {
+                mWaitRemotePeerPromise.set_exception(
+                    std::make_exception_ptr(std::runtime_error("Timeout to find the remote peer")));
+                break;
+            }
+            ping();
+            sleep_for(100ms);
+        }
+        wait_remote_peer_future.get();
+
+        std::future<void> dc_future = mDataChannelPromise.get_future();
+        mPeerConnection = initiatePeerConnection();
+        string label = _data_channel_label.get();
+        mDataChannel = mPeerConnection->createDataChannel(label);
+        registerDataChannelCallBacks(mDataChannel);
+        dc_future.get();
+    }
+    else
+    {
+        std::future<void> dc_future = mDataChannelPromise.get_future();
+        base::Time start_time = base::Time::now();
+        // Wait for the datachannel open until the timeout
```

Comment:
There is a way to wait for a future "until a certain time", we saw that together in one of our calls. Use this, which will avoid the need to have two different branches.

Reply:
https://cplusplus.com/reference/future/future/wait_until/

Reply:
I got confuse with the other one (wait_for)...which is no good for our use case. That other one fits, thanks!